src/state/slices/foxEthSlice/foxEthSlice.ts

```diff
@@ -0,0 +1,400 @@
+import type { PayloadAction } from '@reduxjs/toolkit'
+import { createSlice } from '@reduxjs/toolkit'
+import { createApi } from '@reduxjs/toolkit/dist/query/react'
+import { ethAssetId, foxAssetId } from '@shapeshiftoss/caip'
+import { HistoryTimeframe } from '@shapeshiftoss/types'
+import { Fetcher, Token } from '@uniswap/sdk'
+import IUniswapV2Pair from '@uniswap/v2-core/build/IUniswapV2Pair.json'
+import dayjs from 'dayjs'
+import type { EarnOpportunityType } from 'features/defi/helpers/normalizeOpportunity'
+import {
+  foxEthLpAssetId,
+  UNISWAP_V2_WETH_FOX_POOL_ADDRESS,
+} from 'features/defi/providers/fox-eth-lp/constants'
+import farmAbi from 'features/defi/providers/fox-farming/abis/farmingAbi.json'
+import {
+  calculateAPRFromToken0,
+  getEthersProvider,
+  makeTotalLpApr,
+  rewardRatePerToken,
+} from 'plugins/foxPage/utils'
+import { bn, bnOrZero } from 'lib/bignumber/bignumber'
+import { logger } from 'lib/logger'
+import { BASE_RTK_CREATE_API_CONFIG } from 'state/apis/const'
+import { marketData } from 'state/slices/marketDataSlice/marketDataSlice'
+
+import { FOX_TOKEN_CONTRACT_ADDRESS, WETH_TOKEN_CONTRACT_ADDRESS } from './constants'
+import { getOrCreateContract } from './contractManager'
+import type { FoxEthLpEarnOpportunityType, FoxFarmingEarnOpportunityType } from './foxEthCommon'
+import { farmingOpportunities, lpOpportunity } from './foxEthCommon'
+import { fetchPairData } from './utils'
+
+type FoxEthOpportunities = {
+  farmingOpportunities: FoxFarmingEarnOpportunityType[]
+  lpOpportunity: FoxEthLpEarnOpportunityType
+}
+
+type FoxEthState = Record<
+  string, // accountAddress,
+  FoxEthOpportunities
+>
+
+const initialState: FoxEthState = {}
+
+const moduleLogger = logger.child({ namespace: ['foxEthSlice'] })
+
+export const foxEth = createSlice({
+  name: 'foxEth',
+  initialState,
+  reducers: {
+    clear: () => initialState,
+    upsertLpOpportunity: (state, action: PayloadAction<Partial<FoxEthLpEarnOpportunityType>>) => {
+      if (action.payload.accountAddress && !state[action.payload.accountAddress]) {
+        state[action.payload.accountAddress] = {} as FoxEthOpportunities
+      }
+      state[action.payload.accountAddress ?? ''].lpOpportunity = {
+        ...lpOpportunity, // Common LP properties
+        ...(state[action.payload.accountAddress ?? '']?.lpOpportunity ?? {}),
+        ...action.payload,
+      }
+    },
+    upsertFarmingOpportunity: (
+      state,
+      action: PayloadAction<Partial<FoxFarmingEarnOpportunityType>>,
+    ) => {
+      // TODO: This is absolute immer madness 🤮 clean me before opening the PR
```

Comment:
Kind of looks ugly, but this has to do with the inners of working with immer, and inserting non-yet initialized deep properties.
Happy to reimplement it in a different way, leaving this comment for discussion but if we're happy with it I'll revert it

https://immerjs.github.io/immer/pitfalls/